File: server/internal/handlers/tasks.go
package handlers

import (
	"fmt"
	"github.com/gin-gonic/gin"
	"github.com/jinzhu/copier"
	"net/http"
	"server/internal/commons/bizErr"
	"server/internal/models/entities"
	"server/internal/models/requests"
	"server/internal/models/responses"
	"server/internal/services"
	"server/internal/utils"
	"strconv"
	"strings"
	"time"
)

type TasksHandler struct {
	tasksService *services.TasksService
}

func NewTasksHandler(tasksService *services.TasksService) *TasksHandler {
	return &TasksHandler{
		tasksService: tasksService,
	}
}

func (h *TasksHandler) GetTask(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	var resp responses.BaseResponse
	if err != nil {
		resp.ErrorResponse(c, http.StatusBadRequest, bizErr.ParseParamsErr.Error())
		return
	}

	task, err := h.tasksService.GetTask(id)
	if err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}

	data := responses.GetTaskResponse{}
	_ = copier.Copy(&data, &task)
	resp.SuccessResponse(c, http.StatusOK, data)
}

func (h *TasksHandler) ExportTasks(c *gin.Context) {
	status := strings.TrimSpace(c.Query("status"))
	series := strings.TrimSpace(c.Query("series"))
	filePath, err := h.tasksService.ExportTasks(series, status)
	var resp responses.BaseResponse
	if err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	c.FileAttachment(filePath, fmt.Sprintf("tasks_%s.csv", time.Now().String()))
}

func (h *TasksHandler) CreateTask(c *gin.Context) {
	var req requests.CreateTaskRequest
	if err := utils.Bind(c, &req); err != nil {
		return
	}
	task := entities.Task{
		Series: req.Series,
		Model:  "Sybil",
	}

	var resp responses.BaseResponse
	newTask, err := h.tasksService.Create(&task)
	if err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	data := responses.CreateTaskResponse{}
	_ = copier.Copy(&data, &newTask)
	resp.SuccessResponse(c, http.StatusOK, data)
}

func (h *TasksHandler) UpdateTask(c *gin.Context) {
	var req requests.UpdateTaskRequest
	if err := utils.Bind(c, &req); err != nil {
		return
	}
	var task entities.Task
	_ = copier.Copy(&task, req)

	var resp responses.BaseResponse
	if err := h.tasksService.Update(&task); err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.SuccessResponse(c, http.StatusOK, "success")
}

func (h *TasksHandler) PrioritizeTask(c *gin.Context) {
	var req requests.UpdateTaskRequest
	if err := utils.Bind(c, &req); err != nil {
		return
	}
	var task entities.Task
	_ = copier.Copy(&task, req)

	var resp responses.BaseResponse
	if err := h.tasksService.Prioritize(&task); err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.SuccessResponse(c, http.StatusOK, "success")
}

func (h *TasksHandler) GetListPagination(c *gin.Context) {
	pageParam := c.Query("page")
	pageSizeParam := c.Query("pageSize")
	status := strings.TrimSpace(c.Query("status"))
	series := strings.TrimSpace(c.Query("series"))
	sortKey := strings.TrimSpace(c.Query("sortKey"))
	sortOrder := strings.ToLower(strings.TrimSpace(c.Query("sortOrder")))

	var resp responses.BaseResponse
	if pageParam == "" || pageSizeParam == "" {
		resp.ErrorResponse(c, http.StatusBadRequest, bizErr.ParseParamsErr.Error())
		return
	}
	page, _ := strconv.Atoi(pageParam)
	pageSize, _ := strconv.Atoi(pageSizeParam)

	tasks, total, err := h.tasksService.GetListPagination(page, pageSize, status, series, sortKey, sortOrder)
	if err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.SuccessResponse(c, http.StatusOK, responses.GetTaskLists{
		Total: total,
		Tasks: tasks,
	})
}

func (h *TasksHandler) DeleteTask(c *gin.Context) {
	var req requests.DeleteTaskRequest
	if err := utils.Bind(c, &req); err != nil {
		return
	}
	var task entities.Task
	_ = copier.Copy(&task, req)

	var resp responses.BaseResponse
	if err := h.tasksService.Delete(&task); err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.SuccessResponse(c, http.StatusOK, "success")
}

func (h *TasksHandler) GetUploadUrl(c *gin.Context) {
	series := c.Query("series")
	var resp responses.BaseResponse
	if series == "" {
		resp.ErrorResponse(c, http.StatusBadRequest, bizErr.ParseParamsErr)
		return
	}
	url, form, err := h.tasksService.GetUploadUrl(series)
	if err != nil {
		resp.ErrorResponse(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.SuccessResponse(c, http.StatusOK, responses.GetUploadUrlResponse{
		Url:  url,
		Form: form,
	})
}

func (h *TasksHandler) SetWorkerDevice(c *gin.Context) {
	h.tasksService.SetWorkerDevice(c.Query("device") == "cpu")
	var resp responses.BaseResponse
	resp.SuccessResponse(c, http.StatusOK, "success")
}

File: client/src/apis/tasks.ts
import type {
    CreateTaskRequest,
    CreateTaskResponse,
    DeleteTaskRequest,
    GetTaskLists,
    GetTaskResponse,
    GetUploadUrlResponse,
    UpdateTaskRequest,
} from '@/types'
import { http } from '@/apis/common.ts'

export const getTask = (id: string) => {
    return http<GetTaskResponse>(`/tasks/${id}`)
}

export const createTask = (data: CreateTaskRequest) => {
    return http<CreateTaskResponse>('/tasks/create', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { 'Content-Type': 'application/json' },
    })
}

export const updateTask = (data: UpdateTaskRequest) => {
    return http<string>('/tasks/update', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { 'Content-Type': 'application/json' },
    })
}

export const prioritizeTask = (data: UpdateTaskRequest) => {
    return http<string>('/tasks/prioritize', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { 'Content-Type': 'application/json' },
    })
}

export const deleteTask = (data: DeleteTaskRequest) => {
    return http<string>('/tasks/delete', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { 'Content-Type': 'application/json' },
    })
}

export const getUploadUrl = (params: { series: string }) => {
    const query = new URLSearchParams(params).toString()
    return http<GetUploadUrlResponse>(`/tasks/upload?${query}`)
}

export const getTaskList = ({
    page,
    pageSize,
    status,
    series,
    sortKey,
    sortOrder,
}: {
    page: number
    pageSize: number
    status?: number
    series?: string
    sortKey?: string
    sortOrder?: string
}) => {
    const params: Record<string, string> = {
        page: page.toString(),
        pageSize: pageSize.toString(),
    }

    if (status != null) params.status = status.toString()
    if (series && series != '') params.series = series.trim()
    if (sortKey) params.sortKey = sortKey.trim()
    if (sortOrder) params.sortOrder = sortOrder.trim()

    return http<GetTaskLists>(`/tasks/list?${new URLSearchParams(params)}`)
}

export const setWorkerDevice = (params: { device: string }) => {
    const query = new URLSearchParams(params).toString()
    return http<string>(`/tasks/device?${query}`)
}